File: task15/main.py
# Вариант с файлом:
# with open("input.txt") as f:
#     lines = f.read().splitlines()
#     text = lines[0]
#     pattern = lines[1]

def compute_prefix(pattern):
    m = len(pattern)
    d = [0] * m
    d[0] = -1
    k = -1

    for i in range(1, m):
        while k >= 0 and pattern[k + 1] != pattern[i]:
            k = d[k]
        if pattern[k + 1] == pattern[i]:
            k += 1
        d[i] = k

    return d

def kmp_search(text, pattern):
    n = len(text)
    m = len(pattern)
    d = compute_prefix(pattern)
    results = []
    k = -1

    for i in range(n):
        while k >= 0 and pattern[k + 1] != text[i]:
            k = d[k]
        if pattern[k + 1] == text[i]:
            k += 1
        if k == m - 1:
            results.append(i - m + 1)
            k = d[k]

    return results

text = input("Введите текст: ")
pattern = input("Введите образец: ")
print("Вхождения на позициях:", kmp_search(text, pattern))

File: task15/main.cpp
#include <iostream>
#include <vector>
#include <string>
// Вариант с файлом:
// #include <fstream>
using namespace std;

vector<int> computePrefix(string& pattern) {
    int m = pattern.size();
    vector<int> d(m, 0);
    d[0] = -1;
    int k = -1;

    for (int i = 1; i < m; i++) {
        while (k >= 0 && pattern[k + 1] != pattern[i])
            k = d[k];
        if (pattern[k + 1] == pattern[i])
            k++;
        d[i] = k;
    }

    return d;
}

vector<int> kmpSearch(string& text, string& pattern) {
    int n = text.size();
    int m = pattern.size();
    vector<int> d = computePrefix(pattern);
    vector<int> results;
    int k = -1;

    for (int i = 0; i < n; i++) {
        while (k >= 0 && pattern[k + 1] != text[i])
            k = d[k];
        if (pattern[k + 1] == text[i])
            k++;
        if (k == m - 1) {
            results.push_back(i - m + 1);
            k = d[k];
        }
    }

    return results;
}

int main() {
    // Вариант с файлом:
    // ifstream f("input.txt");
    // string text, pattern;
    // getline(f, text);
    // getline(f, pattern);

    string text, pattern;
    cout << "Введите текст: ";
    getline(cin, text);
    cout << "Введите образец: ";
    getline(cin, pattern);

    vector<int> positions = kmpSearch(text, pattern);

    cout << "Вхождения на позициях: ";
    for (int pos : positions)
        cout << pos << " ";

    return 0;
}